File: app/src/main/java/com/wilco375/roosternotification/ScheduleHandler.java
package com.wilco375.roosternotification;

import android.content.Context;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.ArrayList;
import java.util.List;

public class ScheduleHandler {
    public static Schedule[] getSchedule(Context context){
        return readSchedule(context);
    }

    public static void setSchedule(Context context, Schedule[] schedule){
        writeSchedule(context,schedule);
    }

    public static Schedule[] getScheduleByDay(Context context, int day){
        Schedule[] schedule = readSchedule(context);
        List<Schedule> scheduleByDay = new ArrayList<>();
        for(Schedule scheduleItem : schedule){
            if(scheduleItem.getDay() == day){
                scheduleByDay.add(scheduleItem);
            }
        }
        return Utils.scheduleListToArray(scheduleByDay);
    }

    private static void writeSchedule(Context context, Schedule[] schedule){
        try {
            File file = new File(context.getFilesDir()+"/schedule");
            if(file.exists()) file.delete();
            file.createNewFile();
            FileOutputStream fos = new FileOutputStream(file);
            ObjectOutputStream oos = new ObjectOutputStream(fos);
            oos.writeObject(schedule);
            oos.close();
        }catch (IOException e){
            e.printStackTrace();
        }
    }

    private static Schedule[] readSchedule(Context context){
        try {
            File file = new File(context.getFilesDir() + "/schedule");
            if (!file.exists()) return new Schedule[0];
            FileInputStream fis = new FileInputStream(file);
            ObjectInputStream ois = new ObjectInputStream(fis);
            return (Schedule[]) ois.readObject();
        }catch (IOException e){
            e.printStackTrace();
            return new Schedule[0];
        }catch (ClassNotFoundException e){
            e.printStackTrace();
            return new Schedule[0];
        }
    }
}

File: app/src/main/java/com/wilco375/roosternotification/general/ScheduleHandler.kt
package com.wilco375.roosternotification.general

import android.content.Context
import com.wilco375.roosternotification.`object`.Schedule
import java.io.*

object ScheduleHandler {
    /**
     * Save the schedule
     */
    fun getSchedule(context: Context): Schedule {
        return readSchedule(context)
    }

    /**
     * Update the schedule
     */
    fun setSchedule(context: Context, schedule: Schedule) {
        writeSchedule(context, schedule)
    }

    /**
     * Write schedule object to storage
     */
    private fun writeSchedule(context: Context, schedule: Schedule) {
        // Don't update the schedule if the new one is empty
        if(schedule.getAllScheduleItems().isEmpty()) return

        try {
            val file = File("${context.filesDir}/schedule")
            if (file.exists() && file.delete() && file.createNewFile()) {
                val fos = FileOutputStream(file)
                val oos = ObjectOutputStream(fos)
                oos.writeObject(schedule)
                oos.close()
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    /**
     * Get schedule object from storage
     */
    private fun readSchedule(context: Context): Schedule {
        try {
            val file = File("${context.filesDir}/schedule")
            if (!file.exists()) return Schedule()
            val fis = FileInputStream(file)
            val ois = ObjectInputStream(fis)
            val obj = ois.readObject()
            if (obj is Schedule) {
                return obj
            }
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        }

        return Schedule()
    }
}
